src/Framework/Support/Facades/Facade.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace Give\Framework\Support\Facades;
+
+use Give\Framework\Exceptions\Primitives\Exception;
+
+/**
+ * Class Facade
+ *
+ * This class provides a way of taking a normal instance class and creating a static facade out of it. It does it in
+ * such a way, though, that the facade is still mockable. It does this by instantiating the decorated class through the
+ * Give Service Container (SC). So by injecting a mock singleton of the decorated class in the SC, it can be mocked.
+ *
+ * To use this, simply make a new facade class which extends this once, then override the getFacadeClass and return the
+ * class to be decorated, for example: return MyClass::class;
+ *
+ * To help the IDE, take the methods from the decorated class and add them your class docblock. So if Repository had an
+ * insert method, you would add "@method static Model insert()" to the top.
+ *
+ * @unreleased
+ */
+class Facade {
+	/**
+	 * Static helper for calling the facade methods
+	 *
+	 * @unreleased
+	 *
+	 * @param string $name
+	 * @param array  $arguments
+	 *
+	 * @return mixed
+	 * @throws Exception
+	 */
+	public static function __callStatic( $name, $arguments ) {
+		$instance = give( static::getFacadeClass() );
+
+		return call_user_func_array( [ $instance, $name ], $arguments );
+	}
+
+	/**
+	 * Each facade will call this method and return the intended class instance
+	 *
+	 * @unreleased
+	 *
+	 * @return string
+	 * @throws Exception
+	 */
+	protected static function getFacadeClass() {
+		throw new Exception( 'This method must be overridden and return the class to decorate' );
+	}
+}
```

Comment:
@JasonTheAdams is there any reason we couldn't make this an abstract class/method so this is forced to be used or does _static_ make it problematic?

Reply:
There is no such thing as an `abstract static` method, I'm afraid. We'd have to change how this works like so:
1. The `Facade::__callStatic` method is fired upon calling some method
2. The Facade class instantiates itself
3. The Facade uses an abstract instance method to get the class to be used
4. The class method is called

This is more closely to how Laravel works, but I opted for this direction for two reasons:
- To avoid potential method naming collision with the instance class
- To avoid the need to instantiate the Facade class, making two instances per Static Facade pair (that said, it's _very_ lightweight).

That said, I'm definitely open to it if you can think of a major disadvantage, here. The biggest that I can see — which puts me on the edge — is that this would not be caught by static analysis, wherein an abstract method would.

Reply:
@JasonTheAdams gotchya, yeah honestly the only thing I was concerned about with the current direction is not having static analysis for the required method.

Reply:
Took at stab at this in 7d438eb